src/protocols/musig2/mod.rs

```diff
@@ -0,0 +1,167 @@
+#![allow(non_snake_case)]
+
+//! Simple ed25519
+//!
+//! See https://tools.ietf.org/html/rfc8032
+//! This is an implementation of the Musig2 protocol as shown in https://eprint.iacr.org/2020/1261.pdf with the addition named Musig2* suggested in Section B of the paper.
+//! We implement the v = 2 (NUMBER_OF_NONCES) version, meaning there are 2 nonces generated by each party.
+
+use super::{ExpandedKeyPair, Signature};
+use curv::cryptographic_primitives::hashing::DigestExt;
+use curv::elliptic::curves::{Ed25519, Point, Scalar};
+use rand::Rng;
+use sha2::{digest::Digest, Sha512};
+
+const NUMBER_OF_NONCES: usize = 2;
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct PublicKeyAgg {
+    pub agg_public_key: Point<Ed25519>,
+    pub musig_coefficient: Scalar<Ed25519>,
+}
+
+impl PublicKeyAgg {
+    #[allow(clippy::unnecessary_sort_by)]
+    pub fn key_aggregation_n(
+        mut public_keys: Vec<Point<Ed25519>>,
+        my_public_key: &Point<Ed25519>,
+    ) -> PublicKeyAgg {
+        let mut my_coeff = Scalar::zero();
```

Comment:
We should probably make this an Option and make sure that at the end it is `Some` otherwise it could be that the user messed up and his public key is not inside the public_keys list at all